File: evd_unity_sim_app/Assets/Application/_Scripts/EvD Script Model/data/Region.cs
using System.Collections;
using System.Collections.Generic;


namespace EvD
{
    namespace Data
    {

        [System.Serializable]
        public class Region : Pose
        {
            /*
             * Class Constants
             */

            public static readonly float DEFAULT_ORIENTATION_LIMIT = 1.0f;

            /*
             * Private Members
             */
            
            private bool _freeOrientation;
            private float _uncertaintyOrientationLimit;
            private Orientation _uncertaintyOrientationAltTarget = null;

            /* 
             * Constructors
             */

            public Region(Position centerPosition = null, Orientation centerOrientation = null, bool freeOrientation = true,
                          float uncertaintyOrientationLimit = 1.0f, Orientation uncertaintyOrientationAltTarget = null,
                          string type = "", string name = "", string uuid = null, Node parent = null, bool appendType = true)
            : base(centerPosition,centerOrientation, appendType ? "region." + type : type, name,uuid,parent,appendType)
            {
                this.freeOrientation = freeOrientation;
                this.uncertaintyOrientationLimit = uncertaintyOrientationLimit;
                this.uncertaintyOrientationAltTarget = uncertaintyOrientationAltTarget;
            }

            public static Region FromDct(Dictionary<string,object> dct)
            {
                return new Region(
                    centerPosition: Position.FromDict((Dictionary<string,object>)dct["center_position"]),
                    centerOrientation: Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]),
                    freeOrientation: (bool)dct["free_orientation"],
                    uncertaintyOrientationLimit: (float)dct["uncertainty_orientation_limit"],
                    uncertaintyOrientationAltTarget: Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]),
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string,object> ToDict() 
            {
                var dct = base.ToDict();
                dct.Remove("position");
                dct.Remove("orientation");
                dct.Add("center_position",position.ToDict());
                dct.Add("center_orientation",orientation.ToDict());
                dct.Add("free_orientation",freeOrientation);
                dct.Add("uncertainty_orientation_limit",uncertaintyOrientationLimit);
                dct.Add("uncertainty_orientation_alt_target",uncertaintyOrientationAltTarget.ToDict());
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public Position centerPosition
            {
                get
                {
                    return this.position;
                }

                set
                {
                    this.position = value;
                    UpdatedAttribute("center_position","set");
                }
            }

            public Orientation centerOrientation
            {
                get
                {
                    return this.orientation;
                }

                set
                {
                    this.orientation = value;
                    UpdatedAttribute("center_orientation","set");
                }
            }

            public bool freeOrientation
            {
                get
                {
                    return _freeOrientation;
                }

                set
                {
                    if (_freeOrientation != value) 
                    {
                        _freeOrientation = value;
                        UpdatedAttribute("free_orientation","set");
                    }
                }
            }

            public float uncertaintyOrientationLimit
            {
                get
                {
                    return _uncertaintyOrientationLimit;
                }

                set
                {
                    if (_uncertaintyOrientationLimit != value) 
                    {
                        _uncertaintyOrientationLimit = value;
                        UpdatedAttribute("uncertainty_orientation_limit","set");
                    }
                }
            }

            public Orientation uncertaintyOrientationAltTarget
            {
                get 
                {
                    return _uncertaintyOrientationAltTarget;
                }

                set
                {
                    if (_uncertaintyOrientationAltTarget != value) 
                    {
                        if (_uncertaintyOrientationAltTarget != null)
                        {
                            _uncertaintyOrientationAltTarget.RemoveFromCache();
                        }

                        _uncertaintyOrientationAltTarget = value;
                        if (_uncertaintyOrientationAltTarget != null) 
                        {
                            _uncertaintyOrientationAltTarget.parent = this;
                        }
                        
                        UpdatedAttribute("uncertainty_orientation_alt_target","set");
                    }
                }
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("center_position"))
                {
                    centerPosition = Position.FromDict((Dictionary<string,object>)dct["center_position"]);
                }

                if (dct.ContainsKey("center_orientation"))
                {
                    centerOrientation = Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]);
                }

                if (dct.ContainsKey("free_orientation"))
                {
                    freeOrientation = (bool)dct["free_orientation"];
                }

                if (dct.ContainsKey("uncertainty_orientation_limit"))
                {
                    uncertaintyOrientationLimit = (float)dct["uncertainty_orientation_limit"];
                }

                if (dct.ContainsKey("uncertainty_orientation_alt_target"))
                {
                    uncertaintyOrientationAltTarget = Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]);
                }

                base.Set(dct);
            }

            /*
             * Cache Methods
             */

            public override void RemoveFromCache()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.RemoveFromCache();
                }
                
                base.RemoveFromCache();
            }

            public override void AddToCache()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.AddToCache();
                }

                base.AddToCache();
            }

            /*
             * Update Methods
             */

            public override void LateConstructUpdate()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.LateConstructUpdate();
                }

                base.LateConstructUpdate();
            }

            public override void DeepUpdate()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.DeepUpdate();
                }

                base.DeepUpdate();

                UpdatedAttribute("center_position","update");
                UpdatedAttribute("center_orientation","update");
                UpdatedAttribute("free_orientation","update");
                UpdatedAttribute("uncertainty_orientation_limit","update");
                UpdatedAttribute("uncertainty_orientation_alt_target","update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("center_position","update");
                UpdatedAttribute("center_orientation","update");
                UpdatedAttribute("free_orientation","update");
                UpdatedAttribute("uncertainty_orientation_limit","update");
                UpdatedAttribute("uncertainty_orientation_alt_target","update");
            }
        }

        [System.Serializable]
        public class CubeRegion : Region
        {
            /*
            * Private Members
            */

            private float _uncertaintyX;
            private float _uncertaintyY;
            private float _uncertaintyZ;

            /*
            * Constructors
            */

            public CubeRegion(Position centerPosition = null, Orientation centerOrientation = null, float uncertaintyX = 0.1f,
                              float uncertaintyY = 0.1f, float uncertaintyZ = 0.1f, bool freeOrientation = true, 
                              float uncertaintyOrientationLimit = 1.0f, Orientation uncertaintyOrientationAltTarget = null,
                              string type = "", string name = "", string uuid = null, Node parent = null, bool appendType = true)
            : base(centerPosition, centerOrientation, freeOrientation, uncertaintyOrientationLimit, uncertaintyOrientationAltTarget, 
                   appendType ? "cube-region." + type : type, name,uuid,parent,appendType)
            {
                this.uncertaintyX = uncertaintyX;
                this.uncertaintyY = uncertaintyY;
                this.uncertaintyZ = uncertaintyZ;
            }

            public new static CubeRegion FromDict(Dictionary<string,object> dct)
            {
                return new CubeRegion(
                    centerPosition: Position.FromDict((Dictionary<string,object>)dct["center_position"]),
                    centerOrientation: Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]),
                    uncertaintyX: (float)dct["uncertainty_x"],
                    uncertaintyY: (float)dct["uncertainty_y"],
                    uncertaintyZ: (float)dct["uncertainty_z"],
                    freeOrientation: (bool)dct["free_orientation"],
                    uncertaintyOrientationLimit: (float)dct["uncertainty_orientation_limit"],
                    uncertaintyOrientationAltTarget: Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]),
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string, object> ToDict()
            {
                var dct = base.ToDict();
                dct.Add("uncertainty_x",uncertaintyX);
                dct.Add("uncertainty_y",uncertaintyY);
                dct.Add("uncertainty_z",uncertaintyZ);
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public float uncertaintyX
            {
                get
                {
                    return _uncertaintyX;
                }

                set
                {
                    if (_uncertaintyX != value)
                    {
                        _uncertaintyX = value;
                        UpdatedAttribute("uncertainty_x","set");
                    }
                }
            }

            public float uncertaintyY
            {
                get
                {
                    return _uncertaintyY;
                }

                set
                {
                    if (_uncertaintyY != value)
                    {
                        _uncertaintyY = value;
                        UpdatedAttribute("uncertainty_y","set");
                    }
                }
            }

            public float uncertaintyZ
            {
                get
                {
                    return _uncertaintyZ;
                }

                set
                {
                    if (_uncertaintyZ != value)
                    {
                        _uncertaintyZ = value;
                        UpdatedAttribute("uncertainty_z","set");
                    }
                }
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("uncertainty_x"))
                {
                    uncertaintyX = (float)dct["uncertainty_x"];
                }

                if (dct.ContainsKey("uncertainty_y"))
                {
                    uncertaintyY = (float)dct["uncertainty_y"];
                }

                if (dct.ContainsKey("uncertainty_z"))
                {  
                    uncertaintyZ = (float)dct["uncertainty_z"];
                }

                base.Set(dct);
            }

            /*
            * Update Methods
            */

            public override void DeepUpdate()
            {

                base.DeepUpdate();

                UpdatedAttribute("uncertainty_x", "update");
                UpdatedAttribute("uncertainty_y", "update");
                UpdatedAttribute("uncertainty_z", "update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("uncertainty_x", "update");
                UpdatedAttribute("uncertainty_y", "update");
                UpdatedAttribute("uncertainty_z", "update");
            }
        }


        [System.Serializable]
        public class SphereRegion : Region
        {
            /*
            * Private Members
            */

            private float _uncertaintyRadius;

            /*
            * Constructors
            */

            public SphereRegion(Position centerPosition = null, Orientation centerOrientation = null, float uncertaintyRadius = 0.1f,
                              bool freeOrientation = true, float uncertaintyOrientationLimit = 1.0f, Orientation uncertaintyOrientationAltTarget = null,
                              string type = "", string name = "", string uuid = null, Node parent = null, bool appendType = true)
            : base(centerPosition, centerOrientation, freeOrientation, uncertaintyOrientationLimit, uncertaintyOrientationAltTarget, 
                   appendType ? "sphere-region." + type : type, name,uuid,parent,appendType)
            {
                this.uncertaintyRadius = uncertaintyRadius;
            }

            public new static SphereRegion FromDict(Dictionary<string,object> dct)
            {
                return new SphereRegion(
                    centerPosition: Position.FromDict((Dictionary<string,object>)dct["center_position"]),
                    centerOrientation: Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]),
                    uncertaintyRadius: (float)dct["uncertainty_radius"],
                    freeOrientation: (bool)dct["free_orientation"],
                    uncertaintyOrientationLimit: (float)dct["uncertainty_orientation_limit"],
                    uncertaintyOrientationAltTarget: Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]),
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string, object> ToDict()
            {
                var dct = base.ToDict();
                dct.Add("uncertainty_radius",uncertaintyRadius);
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public float uncertaintyRadius
            {
                get
                {
                    return _uncertaintyRadius;
                }

                set
                {
                    if (_uncertaintyRadius != value)
                    {
                        _uncertaintyRadius = value;
                        UpdatedAttribute("uncertainty_radius","set");
                    }
                }
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("uncertainty_radius"))
                {
                    uncertaintyRadius = (float)dct["uncertainty_radius"];
                }

                base.Set(dct);
            }

            /*
            * Update Methods
            */

            public override void DeepUpdate()
            {

                base.DeepUpdate();

                UpdatedAttribute("uncertainty_radius", "update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("uncertainty_radius", "update");
            }
        }
    }
}

File: evd_authoring_app/src/model/evd_script/data/region.js
import { Pose, Position, Orientation } from './geometry';


export class Region extends Pose {

    /*
     * Class constants
     */

    static DEFAULT_ORIENTATION_LIMIT = 1;

    /*
    * Data structure methods
    */

    static typeString() {
        return 'region.';
    }

    static fullTypeString() {
        return Pose.fullTypeString() + registerIcons.typeString();
    }

    constructor(centerPosition=null, centerOrientation=null, freeOrientation=true, uncertaintyOrientationLimit=1, uncertaintyOrientationAltTarget=null, type='', name='', uuid=null, parent=null, appendType=true) {
        this._freeOrt = null;
        this._uncertOrtLim = null;
        this._uncertOrtTarget = null;

        super(
            position= centerPosition,
            orientation= centerOrientation,
            type= (appendType) ? 'region.'+type : type,
            name= name,
            uuid= uuid,
            parent= parent,
            appendType= appendType
        );

        this.freeOrientation = freeOrientation;
        this.uncertaintyOrientationLimit = uncertaintyOrientationLimit;
        this.uncertaintyOrientationAltTarget = uncertaintyOrientationAltTarget;
    }

    toDict() {
        let msg = super.toDict();

        delete msg.position;
        delete msg.orientation;

        msg = {
            ...msg,
            center_position: this.position.toDict(),
            center_orientation: this.orientation.toDict(),
            free_orientation: this.freeOrientation,
            uncertainty_orientation_limit: this.uncertaintyOrientationLimit,
            uncertainty_orientation_alt_target: this.uncertaintyOrientationAltTarget
        }
        return msg;
    }

    static fromDict(dct) {
        return Region(
            centerPosition= Position.fromDict(dct.center_position),
            centerOrientation= Orientation.fromDict(dct.center_orientation),
            freeOrientation= dct.free_orientation,
            uncertaintyOrientationLimit= dct.uncertainty_orientation_limit,
            uncertaintyOrientationAltTarget= Orientation.fromDict(dct.uncertainty_orientation_alt_target),
            type= dct.type,
            name= dct.name,
            uuid= dct.uuid,
            appendType= false
        );
    }

    toBlockly() {
        //TOOD implement this
        return {};
    }

    /*
    * Data accessor/modifier methods
    */

    get centerPosition() {
        return this.position;
    }

    set centerOrientation(value) {
        this.position = value;
        this.updatedAttribute('center_position','set');
    }

    get centerOrientation() {
        return this.orientation;
    }

    set centerOrientation(value) {
        this.orientation = value;
        this.updatedAttribute('center_orientation','set');
    }

    get freeOrientation() {
        return self._freeOrt;
    }

    set freeOrientation(value) {
        // Note this will destroy previous uncertainty values

        if (this._freeOrt !== value) {
            if (! value instanceof Boolean) {
                throw new Error('Value must be a boolean');
            }

            this._freeOrt = value;
            if (value) {
                this.uncertaintyOrientationLimit = null;
                this.uncertaintyOrientationAltTarget = null;
            } else {
                this.uncertaintyOrientationLimit = Region.DEFAULT_ORIENTATION_LIMIT;
                this.uncertaintyOrientationAltTarget = null;
            }
        }

        this.updatedAttribute('free_orientation','set');
    }

    get uncertaintyOrientationLimit() {
        return this._uncertOrtLim;
    }

    set uncertaintyOrientationLimit(value) {
        if (this._uncertOrtLim !== value) {
            if (this._freeOrt) {
                throw new Error('Orientation is set as free, change freeOrientation to false first!');
            }

            if (value < 0) {
                throw new Error('Value must be positive or zero');
            }

            this._uncertOrtLim = value;
            this.updatedAttribute('uncertainty_orientation_limit','set');
        }
    }

    get uncertaintyOrientationAltTarget() {
        return this._uncertOrtTarget;
    }

    set uncertaintyOrientationAltTarget(value) {
        if (this._uncertOrtTarget !== value) {
            if (this._freeOrt) {
                throw new Error('Orientation is set as free, change freeOrientation to false first!');
            }

            if (this._uncertOrtTarget !== null) {
                this._uncertOrtTarget.removeFromCache();
            }

            this._uncertOrtTarget = value;
            if (this._uncertOrtTarget !== null) {
                this._uncertOrtTarget.parent = this;
            }

            this.updatedAttribute('uncertainty_orientation_alt_target','set');
        }
    }

    set(dct) {

        if ('center_position' in dct) {
            this.centerPosition = Position.fromDict(dct.center_position);
        }

        if ('center_orientation' in dct) {
            this.centerOrientation = Orientation.fromDict(dct.center_orientation);
        }

        if ('free_orientation' in dct) {
            this.freeOrientation = dct.free_orientation;
        }

        if ('uncertainty_orientation_limit' in dct) {
            this.uncertaintyOrientationLimit = dct.uncertainty_orientation_limit;
        }

        if ('uncertainty_orientation_alt_target' in dct) {
            this.uncertaintyOrientationAltTarget = Orientation.fromDict(dct.uncertainty_orientation_alt_target);
        }

        super.set(dct);
    }

    /*
    * Cache Methods 
    */

    removeFromCache() {
        if (this.uncertaintyOrientationAltTarget !== null) {
            this.uncertaintyOrientationAltTarget.removeFromCache();
        }

        super.removeFromCache();
    }

    addToCache() {
        if (this.uncertaintyOrientationAltTarget != null) {
            this.uncertaintyOrientationAltTarget.addToCache();
        }

        super.addToCache();
    }

    /*
    * Update methods
    */

    lateConstructUpdate() {
        if (this.uncertaintyOrientationAltTarget !== null) {
            this.uncertaintyOrientationAltTarget.lateConstructUpdate();
        }

        super.lateConstructUpdate();
    }

    deepUpdate() {

        if (this.uncertaintyOrientationAltTarget !== null) {
            this.uncertaintyOrientationAltTarget.deepUpdate();
        }

        super.deepUpdate();

        this.updatedAttribute('center_position','update');
        this.updatedAttribute('center_orientation','update');
        this.updatedAttribute('free_orientation','update');
        this.updatedAttribute('uncertainty_orientation_limit','update');
        this.updatedAttribute('uncertainty_orientation_alt_target','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('center_position','update');
        this.updatedAttribute('center_orientation','update');
        this.updatedAttribute('free_orientation','update');
        this.updatedAttribute('uncertainty_orientation_limit','update');
        this.updatedAttribute('uncertainty_orientation_alt_target','update');
    }
}

export class CubeRegion extends Region {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'cube-region.';
    }

    static fullTypeString() {
        return Region.fullTypeString() + CubeRegion.type();
    }

    constructor(centerPosition=null, centerOrientation=null, uncertaintyX=1, uncertaintyY=0, uncertaintyZ=0, freeOrientation=true, uncertaintyOrientationLimit=1, uncertaintyOrientationAltTarget=null, type='', name='', uuid=null, parent=null, appendType=true) {
        this._uncertaintyX = null;
        this._uncertaintyY = null;
        this._uncertaintyZ = null;

        super(
            centerPosition= centerPosition,
            centerOrientation= centerOrientation,
            freeOrientation= freeOrientation,
            uncertaintyOrientationLimit= uncertaintyOrientationLimit,
            uncertaintyOrientationAltTarget= uncertaintyOrientationAltTarget,
            type= (appendType) ? 'cube-region.'+type : type,
            name= name,
            uuid= uuid,
            parent= parent,
            appendType= appendType
        );

        this.uncertaintyX = uncertaintyX;
        this.uncertaintyY = uncertaintyY;
        this.uncertaintyZ = uncertaintyZ;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            uncertainty_x: this.uncertaintyX,
            uncertainty_y: this.uncertaintyY,
            uncertainty_z: this.uncertaintyZ
        };
        return msg;
    }

    static fromDict(dct) {
        return CubeRegion(
            centerPosition= dct.center_position,
            centerOrientation= dct.center_orientation,
            uncertaintyX= dct.uncertainty_x,
            uncertaintyY= dct.uncertainty_y,
            uncertaintyZ= dct.uncertainty_z,
            freeOrientation= dct.free_orientation,
            uncertaintyOrientationLimit= dct.uncertainty_orientation_limit,
            uncertaintyOrientationAltTarget= dct.uncertainty_orientation_alt_target,
            type= dct.type,
            name= dct.name,
            uuid= dct.uuid,
            appendType= false
        );
    }

    toBlockly() {
        // TODO implement this
        return {};
    }

    /* 
    * Data accessor/modifier methods
    */

    get uncertaintyX() {
        return this._uncertaintyX;
    }

    set uncertaintyX(value) {
        if (this._uncertaintyX !== value) {
            this._uncertaintyX = value;
            this.updatedAttribute('uncertainty_x','set');
        }
    }

    get uncertaintyY() {
        return this._uncertaintyY;
    }

    set uncertaintyY(value) {
        if (this._uncertaintyY !== value) {
            this._uncertaintyY = value;
            this.updatedAttribute('uncertainty_y','set');
        }
    }

    get uncertaintyZ() {
        return this._uncertaintyZ;
    }

    set uncertaintyZ(value) {
        if (this._uncertaintyZ !== value) {
            this._uncertaintyZ = value;
            this.updatedAttribute('uncertainty_z','set');
        }
    }

    set(dct) {

        if ('uncertainty_x' in dct) {
            this.uncertaintyX = dct.uncertainty_x;
        }   

        if ('uncertainty_y' in dct) {
            this.uncertaintyY = dct.uncertainty_y;
        }

        if ('uncertainty_z' in dct) {
            this.uncertaintyZ = dct.uncertainty_z;
        }

        super.set(dct);
    }

    /*
    * Update methods
    */

    deepUpdate() {
        super.deepUpdate();

        this.updatedAttribute('uncertainty_x','update');
        this.updatedAttribute('uncertainty_y','update');
        this.updatedAttribute('uncertainty_z','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('uncertainty_x','update');
        this.updatedAttribute('uncertainty_y','update');
        this.updatedAttribute('uncertainty_z','update');
    }
}

export class SphereRegion extends Region {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'sphere-region.';
    }

    static fullTypeString() {
        return Region.fullTypeString() + SphereRegion.type();
    }

    constructor(centerPosition=null, centerOrientation=null, uncertaintyRadius=1, freeOrientation=true, uncertaintyOrientationLimit=1, uncertaintyOrientationAltTarget=null, type='', name='', uuid=null, parent=null, appendType=true) {
        this._uncertaintyRadius = null;

        super(
            centerPosition= centerPosition,
            centerOrientation= centerOrientation,
            freeOrientation= freeOrientation,
            uncertaintyOrientationLimit= uncertaintyOrientationLimit,
            uncertaintyOrientationAltTarget= uncertaintyOrientationAltTarget,
            type= (appendType) ? 'sphere-region.'+type : type,
            name= name,
            uuid= uuid,
            parent= parent,
            appendType= appendType
        );

        this.uncertaintyRadius = uncertaintyRadius;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            uncertainty_radius: this.uncertaintyRadius,
        };
        return msg;
    }

    static fromDict(dct) {
        return SphereRegion(
            centerPosition= dct.center_position,
            centerOrientation= dct.center_orientation,
            uncertaintyRadius= dct.uncertainty_radius,
            freeOrientation= dct.free_orientation,
            uncertaintyOrientationLimit= dct.uncertainty_orientation_limit,
            uncertaintyOrientationAltTarget= dct.uncertainty_orientation_alt_target,
            type= dct.type,
            name= dct.name,
            uuid= dct.uuid,
            appendType= false
        );
    }

    toBlockly() {
        // TODO implement this
        return {};
    }

    /* 
    * Data accessor/modifier methods
    */

    get uncertaintyRadius() {
        return this._uncertaintyRadius;
    }

    set uncertaintyRadius(value) {
        if (this._uncertaintyRadius !== value) {
            this._uncertaintyRadius = value;
            this.updatedAttribute('uncertainty_radius','set');
        }
    }

    set(dct) {

        if ('uncertainty_radius' in dct) {
            this.uncertaintyRadius = dct.uncertainty_radius;
        }   

        super.set(dct);
    }

    /*
    * Update methods
    */

    deepUpdate() {
        super.deepUpdate();

        this.updatedAttribute('uncertainty_radius','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('uncertainty_radius','update');
    }
}